services/wiki/wiki_path.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package wiki
+
+import (
+	"net/url"
+	"path"
+	"strings"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// To define the wiki related concepts:
+// * Display Segment: the text what user see for a wiki page (aka, the title):
+//   - "Home Page"
+//   - "100% Free"
+//   - "2000-01-02 meeting"
+// * Web Path:
+//   - "/wiki/Home-Page"
+//   - "/wiki/100%25+Free"
+//   - "/wiki/2000-01-02+meeting.-"
+//   - If a segment has a suffix "DashMarker(.-)", it means that there is no dash-space conversion for this segment.
+//   - If a WebPath is a "*.md" pattern, then use it directly as GitPath, to make users can access the raw file.
+// * Git Path (only space doesn't need to be escaped):
+//   - "/.wiki.git/Home-Page.md"
+//   - "/.wiki.git/100%25 Free.md"
+//   - "/.wiki.git/2000-01-02 meeting.-.md"
```

Comment:
Could we have a more suitable method to avoid the link broken between two markdown files in users local editor environment?

Reply:
I don't see there is any "broken". For example?

Reply:
Assume we have 3 markdown files. The two are `2000-01-02 meeting.-.md` and `2000-01-02 meeting.md` which just have a slight difference. Now there is another markdown file which have a link to one of the file. How should the link looks like?

Reply:
That's impossible to resolve. Even in old implement, you can have `a%b.md` and `a%25b.md` together.

So, that's why this PR has this feature:

> Add a "View original git file" link in the Pages list, even if some file names are broken, users still have a chance to edit or remove it, without cloning the wiki repo to local.